src/tests/ftest/control/ms_resilience.py

```diff
@@ -0,0 +1,314 @@
+        if leader not in kill_list:
+            kill_list.pop()
+            kill_list.add(leader)
+        self.log.info("*** stopping leader (%s) + %d others", leader, N-1)
+        self.stop_group_servers(self.server_group, kill_list)
+
+        return kill_list
+
+    def verify_resiliency(self, N):
+        """Verify 2N+1 resiliency
+
+        This method will launch 2 * N + 1 servers, stop the MS leader,
+        remove the leader host from replicas, verify that a new leader
+        has been elected from the remaining hosts in replicas and verify
+        that the MS is stll accessible by creating a new pool.
+        """
+        replicas = self.launch_servers((2 * N) + 1)
+        leader = self.verify_leader(replicas)
+
+        # First, kill the leader plus just enough other replicas to
+        # push up to the edge of quorum loss.
+        kill_list = self.kill_servers(leader, replicas, N)
+
+        # Next, verify that one of the replicas has stepped up as
+        # the new leader.
+        survivors = [x for x in replicas if x not in kill_list]
+        self.verify_leader(survivors)
+        self.get_dmg_command().hostlist = self.hostlist_servers
+
+        # Dump the current system state.
+        self.get_dmg_command().system_query()
+
+        # Finally, verify that quorum has been retained by performing
+        # write operations.
+        self.create_pool()
+        self.pool = None
+
+    def verify_lost_resiliency(self, N):
+        """Test that even with 2N+1 resiliency lost, reads still work.
+
+        This method will launch 2 * N + 1 servers, will use a kill_list to
+        shutdown N + 1 replica hosts, including the MS leader. Stopped
+        hosts will be removed from the replicas and then verify that we can
+        access MS to read by performing a dmg system command. It then tries to
+        unsuccessfully make an update to MS. Lastly, we bring back the killed
+        servers and check that MS is once again available for writing.
+        """
+        replicas = self.launch_servers((2 * N) + 1)
+        leader = self.verify_leader(replicas)
+
+        # First, create a pool.
+        self.create_pool()
+
+        # Next, kill the leader plus enough other replicas to
+        # lose quorum.
+        kill_list = self.kill_servers(leader, replicas, N+1)
+
+        self.get_dmg_command().hostlist = self.hostlist_servers
+
+        # Now, try to perform some read-only operations to verify
+        # that they work on a MS without quorum.
+        if not self.get_dmg_command().system_leader_query():
+            self.fail("Can't query system after quorum loss.")
+        if not self.find_pool(self.pool.uuid):
+            self.fail("Can't list pools after quorum loss.")
+        self.pool = None
+
+        # Finally, restart the dead servers and verify that quorum is
+        # regained, which should allow for write operations to succeed again.
+        self.restart_servers(self.server_group, list(kill_list))
```

Comment:
(See other comments):
```suggestion
        self.server_managers[0].restart(list(kill_list))
        self.server_managers[0].verify_expected_states()
```